File: src/eQuantic.UI.Runtime/src/router/navigator.ts
/**
 * The web twin of the C# `Navigator` seam (spec S9): PROGRAMMATIC navigation — the routes a
 * component decides to take (a command palette opening its highlighted row, a redirect after a
 * save) rather than the user clicking a link.
 *
 * A running {@link Router} installs itself here, so programmatic navigation gets the same SPA swap
 * a link click does. With no router — a page that never started one, or one already stopped — it
 * falls back to a full document load, which is the honest behavior for a plain multi-page site.
 */

type NavigationHandler = (destination: string) => void;

let handler: NavigationHandler | null = null;

/** Installed by the Router while it is started (null on stop). */
export function setNavigationHandler(next: NavigationHandler | null): void {
  handler = next;
}

export const Navigator = {
  /** Navigates to `destination` — through the active router, else a full document load. The
   * parameter is named the way the vocabulary names it; `location.assign` takes an href because
   * that is what a browser's own API calls it, which is the one place the word belongs. */
  go(destination: string): void {
    if (!destination) return;
    if (handler) {
      handler(destination);
      return;
    }
    if (typeof window !== 'undefined') window.location.assign(destination);
  },
};

File: src/eQuantic.UI.Primitives/Contracts/Navigator.cs
namespace eQuantic.UI.Primitives;

/// <summary>
/// PROGRAMMATIC navigation — the imperative twin of <see cref="Link"/>, for the routes a component
/// decides to take instead of the user clicking them: a command palette opening its highlighted
/// row, a form redirecting after it saves, a menu action that leaves the page.
/// <para>
/// The vocabulary stays declarative wherever a link fits (a Link is focusable, middle-clickable and
/// crawlable — reach for this only when there is no anchor to click). Each target installs its own
/// <see cref="Handler"/>: the web runtime routes through the SPA router (falling back to a full
/// document load), and <c>PhotonHost</c> installs its navigation seam. With no handler installed
/// the call is a no-op rather than a crash — an SSR pass has nowhere to navigate to.
/// </para>
/// </summary>
public static class Navigator
{
    /// <summary>The active target's navigation seam. One surface owns it (a native host, a browser
    /// document); hosts install it as they start.</summary>
    public static Action<string>? Handler { get; set; }

    /// <summary>
    /// Navigates to <paramref name="destination"/> through the active seam.
    /// <para>
    /// The parameter was called <c>href</c>, which is the DOM's word in the assembly whose rule is
    /// that no name here would exist if the web did not. <c>destination</c> is what
    /// <see cref="Link"/> calls the same thing, and this is its imperative twin — a route is a
    /// route on a target with no anchors at all.
    /// </para>
    /// </summary>
    public static void Go(string destination)
    {
        if (string.IsNullOrEmpty(destination)) return;
        Handler?.Invoke(destination);
    }
}
